### PR TITLE
Adding Farsi/Persian Language

### PR DESCRIPTION
**Pull request by MansourM**
Hey
I started Integrating Farsi/Persian (FA) to scrabble solver:

**Already Done:**
1. basically anywhere locale was used I also added fa_FA locale
2. added Fa locale to sidebar (SVG doesn't show up correctly for some reason)
3. added tile (letter) information
4. added world list, crawler, definition and parser

**TODO:**
1. bugfix
2. parser tests
3. adjust the word-bag count to match with official version
4. translation file
5. RTL support

### CODE REVIEW

File: packages/types/src/Locale.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ enum Locale {
   EN_GB = 'en-GB',
   EN_US = 'en-US',
   ES_ES = 'es-ES',
+  FA_FA = 'fa-Fa',
```
Comment (kamilmielnik) [inline packages/types/src/Locale.ts:7]: Those are IETF language tags, so for Persian used in Iran the code should be `fa-IR`

```suggestion
  FA_IR = 'fa-IR',
```

File: packages/scrabble-solver/src/lib/detectLocale.ts [hunk 2]
```diff
@@ -9,6 +9,11 @@ const detectLocale = (): Locale => {
     return Locale.EN_GB;
   }
 
+  //TODO not sure if this is needed
+  if (window.navigator.languages.includes('fa-FA')) {
```
Comment (kamilmielnik) [inline packages/scrabble-solver/src/lib/detectLocale.ts:13]: Yes, it is needed to detect the language when user opens the app for the first time.

```suggestion
  if (window.navigator.languages.includes('fa') || window.navigator.languages.includes('fa-IR')) {
```

File: packages/scrabble-solver/src/icons/FlagFa.svg [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+<g fill="#da0000" transform="matrix(45,0,0,45,315,180)">
+<g id="emblem_half">
+<path d="M-0.54815,0.83638A0.912046,0.912046 0 0,0 0.328544,-0.722384A1,1 0 0,1 -0.54815,0.83638"/>
+<path d="M0.618339,0.661409A0.763932,0.763932 0 0,0 0.421644,-0.741049A1,1 0 0,1 0.618339,0.661409"/>
+<path d="M0,1 -0.05,0 0,-0.787278A0.309995,0.309995 0 0,0 0.118034,-0.688191V-0.100406L0.077809,0.892905z"/>
+<path d="M-0.02,-0.85 0,-0.831217A0.14431,0.14431 0 0,0 0.252075,-0.967708A0.136408,0.136408 0 0,1 0,-0.924634"/>
+</g>
+<use xlink:href="#emblem_half" transform="scale(-1,1)"/>
+</g>
+</svg>
```
Comment (kamilmielnik) [inline packages/scrabble-solver/src/icons/FlagFa.svg:32]: This version should work well:

```suggestion
<svg xmlns="http://www.w3.org/2000/svg" xmlns:svg="http://www.w3.org/2000/svg" viewBox="0 0 630 360">
  <path style="fill:#da0000" d="M 0,0 H 630 V 360 H 0 Z" />
  <path style="fill:#ffffff" d="M 0,0 H 630 V 240 H 0 Z" />
  <path style="fill:#239f40" d="M 0,0 H 630 V 120 H 0 Z" />
  <g transform="translate(8.4,100.4)">
    <g fill="none" stroke="#fff" stroke-width="2">
      <path d="M0,1H26M1,10V5H9V9H17V5H12M4,9H6M26,9H21V5H29M29,0V9H37V0M33,0V9" transform="scale(1.4)" />
      <path d="M0,7H9M10,7H19" transform="scale(2.8)" />
      <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
      <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
    </g>
    <g>
      <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(56)">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
      <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(112)">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
      <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(168)">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
    </g>
    <g transform="translate(168)">
      <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(56)">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
      <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(112)">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
      <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(168)">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
    </g>
    <g transform="translate(392)">
      <g fill="none" stroke="#ffffff" stroke-width="2">
        <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
        <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
        <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
        <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
      </g>
      <g>
        <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(56)">
          <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
          <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
          <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
          <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
        </g>
        <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(112)">
          <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
          <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
          <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
          <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
        </g>
        <g fill="none" stroke="#ffffff" stroke-width="2" transform="translate(168)">
          <path d="M 0,1 H 26 M 1,10 V 5 h 8 v 4 h 8 V 5 H 12 M 4,9 H 6 M 26,9 H 21 V 5 h 8 m 0,-5 v 9 h 8 V 0 m -4,0 v 9" transform="scale(1.4)" />
          <path d="m 0,7 h 9 m 1,0 h 9" transform="scale(2.8)" />
          <path d="m 0,139.6 h 25.2 m 2.8,0 h 25.2" style="stroke-width:5.6" />
          <path d="m 0,146.6 h 36.4 m -35,12.6 v -7 h 11.2 v 5.6 h 11.2 v -5.6 h -7 m -11.2,5.6 h 2.8 m 28,0 h -7 v -5.6 h 11.2 m 0,-7 v 12.6 h 11.2 v -12.6 m -5.6,0 v 12.6" style="stroke-width:2.8" />
        </g>
      </g>
    </g>
  </g>
  <g fill="#da0000" transform="matrix(45,0,0,45,315,180)">
    <g>
      <path d="M-0.54815,0.83638A0.912046,0.912046 0 0,0 0.328544,-0.722384A1,1 0 0,1 -0.54815,0.83638" />
      <path d="M0.618339,0.661409A0.763932,0.763932 0 0,0 0.421644,-0.741049A1,1 0 0,1 0.618339,0.661409" />
      <path d="M0,1 -0.05,0 0,-0.787278A0.309995,0.309995 0 0,0 0.118034,-0.688191V-0.100406L0.077809,0.892905z" />
      <path d="M-0.02,-0.85 0,-0.831217A0.14431,0.14431 0 0,0 0.252075,-0.967708A0.136408,0.136408 0 0,1 0,-0.924634" />
    </g>
    <g transform="scale(-1,1)">
      <path d="M -0.54815,0.83638 A 0.912046,0.912046 0 0 0 0.328544,-0.722384 1,1 0 0 1 -0.54815,0.83638" />
      <path d="M 0.618339,0.661409 A 0.763932,0.763932 0 0 0 0.421644,-0.741049 1,1 0 0 1 0.618339,0.661409" />
      <path d="M 0,1 -0.05,0 0,-0.787278 a 0.309995,0.309995 0 0 0 0.118034,0.099087 v 0.587785 L 0.077809,0.892905 Z" />
      <path d="M -0.02,-0.85 0,-0.831217 A 0.14431,0.14431 0 0 0 0.252075,-0.967708 0.136408,0.136408 0 0 1 0,-0.924634" />
    </g>
  </g>
</svg>
```

**Review by kamilmielnik — COMMENTED** [review]
Thanks for the PR!
I hope you don't mind some early feedback.

----

About the "Syntax error: misplaced ")"" bug you are having - I still can't reproduce it, everything looks good to me:

image

Something must be wrong with the dictionary file and perhaps it is not being updated due to cache.

Can you remove `~/.scrabble-solver/dictionaries/fa-Fa.txt` and try again?
Then dictionary will be re-fetched.

**Comment by MansourM** [comment]
yes, the bug was a cache problem, fixed after i did a clean build.
two more things i noticed:
1. I just saw that i completely missed the sack part somehow and the letter-sack is currently always empty :)
2. did not test why, but the crawler/parser doesn't work properly, a lot of times i get "Word exists in the dictionary but it does not have a definition." but when i manually check the dictionary website the word exist, you can check 
`https://www.vajehyab.com/moein/%D9%85%D8%AF%D8%B1%D8%B3` 
as an example the Persian word is
`مدرس`

File: packages/scrabble-solver/src/i18n/fa.json [hunk 4]
```diff
@@ -0,0 +1,56 @@
+{
+  "cell.filter-cell": "مقصد - کلیک برای تغییر",
+  "cell.set-blank": "علامت گذاری به عنوان خالی",
+  "cell.set-not-blank": "علامت گذاری به عنوان غیر خالی",
+  "cell.toggle-direction": "جهت تایپ - کلیک برای تغییر",
+  "common.blanks": "خالی",
+  "common.clear": "پاک کردن",
+  "common.close": "بستن",
+  "common.consonants": "حروف صامت",
+  "common.loading": "در حال بارگزاری",
```
Comment (kamilmielnik) [inline packages/scrabble-solver/src/i18n/fa.json:10]: The use of spaces here causes an issue with how this text is displayed:

image

Comment (kamilmielnik) [reply packages/scrabble-solver/src/i18n/fa.json:10]: This patch fixes it:
```patch
diff --git a/packages/scrabble-solver/src/components/Loading/Loading.tsx b/packages/scrabble-solver/src/components/Loading/Loading.tsx
index b76e52c3..903a4ecd 100644
--- a/packages/scrabble-solver/src/components/Loading/Loading.tsx
+++ b/packages/scrabble-solver/src/components/Loading/Loading.tsx
@@ -13,9 +13,15 @@ interface Props {
   wave?: boolean;
 }
 
+const prepareContent = (message: string): string[][] => {
+  const uppercased = message.toLocaleUpperCase();
+  const parts = uppercased.split(' ');
+  return [parts];
+};
+
 const Loading: FunctionComponent<Props> = ({ className, wave = true }) => {
   const translate = useTranslate();
-  const content = useMemo<string[][]>(() => [[translate('common.loading').toUpperCase()]], [translate]);
+  const content = useMemo(() => prepareContent(translate('common.loading')), [translate]);
 
   return (
     <div className={classNames(styles.loading, className)}>
```

image

File: packages/configs/src/scrabble/tiles/tilesFa.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+//FIXME scores are based on persian scrabble wiki but the counts are made up!
```
Comment (kamilmielnik) [inline packages/configs/src/scrabble/tiles/tilesFa.ts:1]: The counts can be found here:
- https://en.wikipedia.org/wiki/Scrabble_letter_distributions#Persian
- http://www.scrabble3d.info/t1811f360-Das-persische-Buchstabenset-von-Scrabble-D.html

Please also see the note from Wikipedia (above link):

> Although Persian letters have up to four forms, Scrabble tiles use the isolated form. The pattern of using the isolated forms in composing words is also found in the Arabic Scrabble set and in Arabic crosswords and is one of the rare situations when Arabic letters are not connected to each other.

Not sure if it already works like this, but I think it should.

**Review by kamilmielnik — COMMENTED** [review]
> I just saw that i completely missed the sack part somehow and the letter-sack is currently always empty :)

`CONSONANTS` and `VOWELS` need to be updated here: https://github.com/kamilmielnik/scrabble-solver/blob/master/packages/constants/src/index.ts#L11-L40


> did not test why, but the crawler/parser doesn't work properly, a lot of times i get "Word exists in the dictionary but it does not have a definition." but when i manually check the dictionary website the word exist, you can check
https://www.vajehyab.com/moein/%D9%85%D8%AF%D8%B1%D8%B3
as an example the Persian word is

[RegExp here](https://github.com/kamilmielnik/scrabble-solver/blob/cd50091/packages/word-definitions/src/lib/normalizeDefinition.ts#L24) will have to be updated to account for Arabic characters. The line should be:

```tsx
// https://stackoverflow.com/a/40732368
const normalizeNonWords: Normalize = (definition) => (/[\w\u0621-\u064A]+/.test(definition) ? definition : '');
```

File: packages/scrabble-solver/src/i18n/fa.json [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  "keyMap.board.toggle-blank": "علامت/عدم علامت گذاری کاشی به عنوان خالی",
+  "keyMap.board.toggle-direction": "تغییر جهت تایپ",
+  "keyMap.board-and-rack": "صفحه و کیسه",
+  "keyMap.board-and-rack.insert-two-letter-tile": "وارد کردن کاشی دو حرفی",
+  "keyMap.board-and-rack.navigate": "حرکت",
+  "keyMap.board-and-rack.remove-tile": "حذف کاشی",
+  "keyMap.board-and-rack.submit": "حل کردن",
+  "keyMap.rack": "کیسه",
+  "keyMap.rack.insert-blank": "وارد کردن کاشی خالی (دکمه اسپیس)",
+  "rack.placeholder": "حروف",
```
[the remark this answers is not in the document]
Comment (MansourM) [reply packages/scrabble-solver/src/i18n/fa.json:38]: yes, ill made the persian equivalent of "letter list" will be 7 chars
`لیست حرف`

File: packages/configs/src/scrabble/tiles/tilesFa.ts (hunk 5 above)
Comment (MansourM) [reply packages/configs/src/scrabble/tiles/tilesFa.ts:1]: nice, although it has separated `ا` and `آ` which are technically the same letter, im also checking with the company who has made the the Persian board game for info, waiting for his reply.

File: packages/scrabble-solver/src/i18n/fa.json (hunk 4 above)
Comment (MansourM) [reply packages/scrabble-solver/src/i18n/fa.json:10]: hmmm, im not very familiar with PR workflow, and i supposed to make this changes and made a new commit or these are already done?

Comment (kamilmielnik) [reply packages/scrabble-solver/src/i18n/fa.json:10]: This needs to be a new commit.
To apply a patch, copy all the code I pasted above, save it as a new file in the repository, call it e.g. `patch.diff`, and then apply the file as a patch: `git apply patch.diff`. Then you can delete the `patch.diff` file and commit the changes.

File: packages/configs/src/scrabble/tiles/tilesFa.ts (hunk 5 above)
Comment (MansourM) [reply packages/configs/src/scrabble/tiles/tilesFa.ts:1]: I got a response, n the physical version  `ا` and `آ` where the same char as I suspected, updated Farsi tileset to match those.

this image is from the boardgame's guidebook:

Screenshot 2023-01-12 224802

File: packages/scrabble-solver/src/i18n/fa.json (hunk 4 above)
Comment (MansourM) [reply packages/scrabble-solver/src/i18n/fa.json:10]: cool, learning new stuff :)
i followed the steps, but I'm getting an error ->  "error: corrupt patch at line 21" if I had to guess maybe its a windows problem.

File: packages/scrabble-solver/src/i18n/fa.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (MansourM) [reply packages/scrabble-solver/src/i18n/fa.json]: constants and vowels can be a bit complicated in Farsi, technically from the letters only `ا`  is a vowel but the thing is `ی`  and `و` can be either based on the word. we also have vowel markings the can be placed on any character like `ب` is `b` but with `َ` marking `بَ` now is read `ba` as in bad, `بِ` is `be` as in bed and `بُ` is `bo` as in bored. and this is not the end :)
so im not really sure how to handle this. my suggestion is to ignore this on Farsi and just put all the letters in a big pile.

**Review by kamilmielnik — APPROVED** [review]
Good job! :1st_place_medal:

File: packages/scrabble-solver/src/i18n/fa.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply packages/scrabble-solver/src/i18n/fa.json]: > my suggestion is to ignore this on Farsi and just put all the letters in a big pile.

Sounds good. Created an issue for it: https://github.com/kamilmielnik/scrabble-solver/issues/174

File: packages/scrabble-solver/src/i18n/fa.json (hunk 4 above)
Comment (kamilmielnik) [reply packages/scrabble-solver/src/i18n/fa.json:10]: Fixed: adafdef1

**Pull request merged**